### PR TITLE
feat: allow users to import from node_modules

### LINKED ISSUE #1008: [Feature Request] extended `import` utilising package manager to provide exportability
**Is your feature request related to a problem? Please describe.**
Extension of the `import` utility  to enable user to import from `node_modules` folders, enabling the ability to extend from external packages. 

**Describe the solution you'd like**
A simple way to extend from `node_modules` folders. From the looks of it, an refactor of 

https://github.com/zenstackhq/zenstack/blob/b4579c7f3a5d47e1f61c5268d3ac8efbebc7557d/packages/schema/src/utils/ast-utils.ts#L96-L106

I assume there would be an additional step somewhere within the language file, and probably within the vscode extension, however haven't yet discovered those locations (will update once I have)

**Describe alternatives you've considered**
Keeping all my files in a single project and making endless copies of them into each individual project. Meaning there's no single source of truth, making it easy for one scheme to fall out of sync with another....

### PR DESCRIPTION
**Pull request by TGTGamer**
Goal: Closes #1008

> [!CAUTION]
> Test for IDE's not completed due to errors running on Windows & Codespace Development Environments.
> ` ERR_PNPM_RECURSIVE_RUN_FIRST_FAIL  jetbrains@1.9.0 build: "./gradlew buildPlugin"`

## Summary
This code defines a function named `resolveImportUri` that takes a `ModelImport` object as input and returns a `URI` object or `undefined`. The function is responsible for resolving the import path of the `ModelImport` object.

## Example Usage
```javascript
const imp: ModelImport = {
  ...
  path: '@eventiva/myModel'
};

const result = resolveImportUri(imp);
console.log(result); // Output: undefined or URI object for "node_modules/@eventiva/myModel.zmodel"
```

### Inputs
- `imp` (type: `ModelImport`): The `ModelImport` object that contains the import path to be resolved.
___
### Flow
1. Check if the `path` property of the `imp` object is undefined, null, or an empty string. If true, return undefined.
2. If the `path` does not end with '.zmodel', append '.zmodel' to the `path`.
3. Check if the `path` is a relative path, an absolute path (Unix), or an absolute path (Windows). If true, skip to step 6.
4. Call the `findNodeModulesFile` function with the `path` as an argument to find the file location in the node_modules directory.
5. If the file location is empty, return undefined.
6. Get the directory URI of the `imp` object using the `getDocument` and `Utils.dirname` functions.
7. Resolve the path using the `Utils.resolvePath` function with the directory URI and the modified `path`.
8. Return the resolved URI.
___
### Outputs
- `URI` object: The resolved URI of the import path.
- `undefined`: If the import path cannot be resolved.
___
### Tests

```cmd
packages/schema test$ ZENSTACK_TEST=1 jest "--silent" "--forceExit"
[76 lines collapsed]
│   machine-id-utils.ts                  |   71.62 |    33.33 |     100 |   71.62 | 25-26,37-41,43-47,54-57,59,65-66,72-73                                                                                          …
│   pkg-utils.ts                         |   45.12 |       50 |   28.57 |   45.12 | 64,68-118,120-134,141-155,157-164                                                                                               …
│   typescript-expression-transformer.ts |   74.12 |    74.54 |      70 |   74.12 | 26-27,80,92,98,103-105,109-110,113,115-118,131-132,138-139,143-144,166-178,209-212,216-218,222-224,228-230,268-269,274-284,304-3…
│   version-utils.ts                     |   84.61 |       50 |     100 |   84.61 | 10-11                                                                                                                           …
│ ---------------------------------------|---------|----------|---------|---------|---------------------------------------------------------------------------------------------------------------------------------…
│ Test Suites: 17 passed, 17 total
│ Tests:       124 passed, 124 total
│ Snapshots:   0 total
│ Time:        534.871 s
└─ Done in 8m 57.3s
```

## Summary by CodeRabbit

- **Refactor**
	- Updated the logic for locating `package.json` files across various modules for improved path resolution.
- **New Features**
	- Enhanced the schema generation process by refining the path resolution logic.
	- Introduced a new method for handling module file searches within node_modules directories.
- **Chores**
	- Added a deprecation notice for the `findPackageJson` function, encouraging the use of the new path searching methods.

### CODE REVIEW

**Comment by TGTGamer** [comment]
@coderabbitai review

**Comment by TGTGamer** [comment]
@ymc9 could you advise where you would like me to write tests in for this, and what specific tests you'd like me to run? Is there a node_module which has a zmodel which we can import as part of the development dependancies or would you like me to initialise a very basic package just to host as that dependancy?

**Comment by ymc9** [comment]
> @ymc9 could you advise where you would like me to write tests in for this, and what specific tests you'd like me to run? Is there a node_module which has a zmodel which we can import as part of the development dependancies or would you like me to initialise a very basic package just to host as that dependancy?

Currently, the only zmodel that resides in node_modules is "@zenstackhq/openapi/plugin.zmodel". The "plugin.zmodel" will be automatically loaded when a plugin is enabled. I believe you can also use it to test the import mechanism without using the plugin. A bit hacky though. Alternatively, you can also just scaffold an npm project in the test case, contrive a dependency, and smuggle a zmodel inside of it.

I think the test cases can belong to "packages/schema/tests/schema/...". What you think?

File: packages/schema/src/utils/pkg-utils.ts [hunk 1]
```diff
@@ -37,6 +37,32 @@ export function findUp<e extends boolean = false>(names: string[], cwd: string =
...
+    if (!name) return undefined;
+    const nodeModules = findUp(['node_modules'], cwd, true);
+    if (!nodeModules) return undefined;
+    const split = name.split('/')
+    const folder = split.slice(0, -1).join('/');
+    const folderLocation = nodeModules.find((dir) =>fs.existsSync(path.join(dir, folder)));
+    if (!folderLocation) return undefined;
+    const fileLocation = path.join(folderLocation, name);
+    return fs.existsSync(fileLocation) ? fileLocation : undefined;
+}
```
[the remark this answers is not in the document]
Comment (TGTGamer) [reply packages/schema/src/utils/pkg-utils.ts:64]: > The function assumes that node_modules directories found by findUp are always valid. Ensure that this assumption holds, especially in complex project structures.

`FindUp` always returns either `string, string[], undefined`. These are handled in the expected manour as per established package conventions. 

> The handling of the file name and folder separation (lines 58-59) is simple but does not account for scoped packages (e.g., @scope/package). Ensure that scoped packages are correctly handled.

The `@scope/package` is handled as npm packages are always created as folders with the `@scope` as the first level folder name. The added `@` makes very little difference when searching for folders.

> The check for file existence (line 63) is appropriate, but consider the performance implications of multiple fs.existsSync calls, especially for large node_modules directories.

This is only called once when the folder is found to ensure that the file exists within the package.

File: packages/schema/src/utils/pkg-utils.ts [hunk 2]
```diff
@@ -37,6 +37,32 @@ export function findUp<e extends boolean = false>(names: string[], cwd: string =
...
+ * If the folder is not found, it returns undefined. 
+ * It then constructs the file location by joining the folder location and the file name with a '.zmodel' extension if the file does not already have an extension. 
+ * Returns the file location if it exists, otherwise returns undefined.
+ * @author Jonathan Stevens (TGTGamer)
+ *
+ * @export
+ * @param name A string representing the name of the file to find in the node_modules folder
+ * @returns Path to a specific file within the node_modules directory
+ */
+export function findNodeModulesFile(name: string, cwd: string = process.cwd()) {
```
Comment (ymc9) [inline packages/schema/src/utils/pkg-utils.ts:54]: Do you think we can actually use `require.resolve` to resolve zmodel? Like `require.resolve('@myorg/mypkg/foo.zmodel')` or `require.resolve('pkg/foo.zmodel')`.

I believe `require.resolve` handles recursive upward look-up internally. Maybe it gets us the same behavior?

**Comment by ymc9** [comment]
Hi @TGTGamer , thanks for continuing to work on it. I've added a few review comments.

**Comment by TGTGamer** [comment]
Changes made. @ymc9 please review :)

File: packages/schema/src/utils/pkg-utils.ts [hunk 3]
```diff
@@ -37,6 +37,31 @@ export function findUp<e extends boolean = false>(names: string[], cwd: string =
     return findUp(names, up, multiple, result);
 }
 
+/**
+ * A function that finds a file within the node_modules directory by the provided name.
```
Comment (ymc9) [inline packages/schema/src/utils/pkg-utils.ts:41]: The documentation is out-dated and needs update.

**Review by ymc9 — DISMISSED** [review]
Looks good now! The only remaining thing is the documentation of `findNodeModulesFile` is outdated. Please help update and the pr is good to merge!

Thank you!

**Review by ymc9 — APPROVED** [review]

**Pull request merged**